fix: delayed suspense causes "flicker"

tjosepo opened the pull request:
<!-- Please use this template for your pull request. -->
<!-- Please use the sections that you need and delete other sections -->

## This PR
<!-- add the description of the PR here -->

- Add an internal `useOpenFeatureClientStatus()` function to make it slightly easier to keep track of the state of the current Client, and know when to suspend or not.

- Replaces the `suspend()` function with a `suspendUntilReady()` function. The `suspendUntilReady()`  function fires immediatly during the first render or when the component cannot render due to the client having changed status.

- Put the Client status and the feature flag details inside React states. This is required to make APIs like `startTransition` work properly with the library

### Related Issues
<!-- add here the GitHub issue that this PR resolves if applicable -->

Fixes #920

### Notes
<!-- any additional notes for this PR -->

### Follow-up Tasks

- [x] Add more tests. Suspense and Transitions are very tricky. While these changes don't seem to have broken any existing tests, we should add new tests to make sure Suspense and Transitions continue to work properly in the future.

- [ ] ~I think there might still be some reactivity issues with some parts of the codebase. The client returned by `useOpenFeatureClient()` didn't seem to update properly when the client changed, which made `useOpenFeatureClientStatus()` not update properly.~ Out of the scope of this PR

### How to test
<!-- if applicable, add testing instructions under this section -->

I built the application with `npm build`, then `npm pack`. I used the modified code on a local project to see if it fixed the issues.

I built projects in CodeSanbox that demonstrate the issues.

For Suspense: https://codesandbox.io/embed/openfeature-suspense-bug-5j7yll?fontsize=14&hidenavigation=1&theme=dark

For Transitions: https://codesandbox.io/embed/openfeature-suspense-bug-forked-lqhyf3?fontsize=14&hidenavigation=1&theme=dark

toddbaert commented:
This looks great!

Don't try to do everything in one go though, I would recommend you just deal with the suspense issues for now, if you want to fix any reactivity issues with the other hooks, try to do that separately.

toddbaert commented:
I also locally built and linked this, and tested it with our[ little demo app](https://github.com/open-feature/react-test-app), and it worked exactly as expected! Please carry on!

Be careful not to leak any new handlers!

Thanks again :pray:

tjosepo commented:
Perfect! I'll try adding a few more tests. I'll try to get the PR ready to review by the end of the day.

toddbaert commented:
> Perfect! I'll try adding a few more tests. I'll try to get the PR ready to review by the end of the day.

No rush!

tjosepo commented:
I modified the Suspense tests to make sure the component suspended immediately when they should.

I also added two tests to make sure transitions work properly (transitions are tricky to get right, so I'm happy tests now cover it!)

toddbaert commented:
@tjosepo Thanks again! I will review this no later than tomorrow.

toddbaert reviewed (approved):
@tjosepo I tried but I can't find anything to complain about :grin:. I also tested it again in our [test app](https://github.com/open-feature/react-test-app). Thanks for identifying this issue and for the fix! :pray: 

Would you like an invite to the OpenFeature org? It comes with no obligation, but allows us to ping you and keep you in the loop on Github. It's the first step on the [contributor ladder](https://github.com/open-feature/community/blob/main/CONTRIBUTOR_LADDER.md).

tjosepo commented:
> Would you like an invite to the OpenFeature org? It comes with no obligation, but allows us to ping you and keep you in the loop on Github. It's the first step on the [contributor ladder](https://github.com/open-feature/community/blob/main/CONTRIBUTOR_LADDER.md).

Sure! We are very likely to be using OpenFeature at my job soon, so I'd love to contribute to it when I can!

lukas-reining reviewed (approved):
Looks really good @tjosepo, thanks! :)

toddbaert commented:
@tjosepo created a PR to add you to the org: https://github.com/open-feature/community/pull/372
